src/execution/join_hashtable.cpp

```diff
@@ -794,4 +888,221 @@ idx_t JoinHashTable::FillWithHTOffsets(data_ptr_t *key_locations, JoinHTScanStat
 	}
 	return key_count;
 }
+
+void JoinHashTable::SwizzleBlocks() {
+	RowDataCollectionScanner::AlignHeapBlocks(*swizzled_block_collection, *swizzled_string_heap, *block_collection,
+	                                          *string_heap, layout);
+}
+
+void JoinHashTable::UnswizzleBlocks() {
+	auto &blocks = swizzled_block_collection->blocks;
+	auto &heap_blocks = swizzled_string_heap->blocks;
+#ifdef DEBUG
+	if (!layout.AllConstant()) {
+		D_ASSERT(blocks.size() == heap_blocks.size());
+		D_ASSERT(swizzled_block_collection->count == swizzled_string_heap->count);
+	}
+#endif
+
+	for (idx_t block_idx = 0; block_idx < blocks.size(); block_idx++) {
+		auto &data_block = blocks[block_idx];
+
+		if (!layout.AllConstant()) {
+			auto block_handle = buffer_manager.Pin(data_block->block);
+
+			auto &heap_block = heap_blocks[block_idx];
+			D_ASSERT(data_block->count == heap_block->count);
+			auto heap_handle = buffer_manager.Pin(heap_block->block);
```

Comment:
It might improve safety here to assert that the heap blocks are pinned?

Reply:
We pin them both here, I don't see why this should be asserted